src/profiles_management/helpers/kfam.py

```diff
@@ -104,3 +217,85 @@ def list_contributor_authorization_policies(
+
+    The user of the resource will be used as a key and its role as the value.
+
+    Args:
+        resources: List of KFAM RoleBindings or AuthorizationPolicies.
+
+    Returns:
+        Dictionary with keys the user names and values the roles, derived from parsing all
+        the provided resources.
+    """
+    contributor_roles_dict = {}
+    for resource in resources:
+        if has_kfam_annotations(resource):
+            user = get_contributor_user(resource)
+            role = get_contributor_role(resource)
+            contributor_roles_dict[user] = contributor_roles_dict.get(user, []) + [role]
+
+    return contributor_roles_dict
+
+
+def delete_rolebindings_not_matching_profile_contributors(
+    client: Client,
+    profile: classes.Profile,
+    existing_rolebindings: List[RoleBinding],
+) -> List[RoleBinding]:
+    """Delete RoleBindings in the cluster that doesn't match Contributors in PMR Profile.
+
+    Args:
+        client: The lightkube client to use.
+        profile: The PMR Profile to create RoleBindings based on its Contributors.
+        existing_rolebindings: RoleBindings in the cluster that will be evaluated for deletion.
+
+    Returns:
+        The remaining resources, after removing the deleted ones from the existing_resources.
+    """
+    role_bindings_to_delete = []
+    remaining_role_bindings = []
+
+    for rb in existing_rolebindings:
+        if not resource_matches_profile_contributor(rb, profile):
+            log.info(
+                "RoleBinding '%s' doesn't belong to Profile. Will delete it.",
+                k8s.get_name(rb),
+            )
+            role_bindings_to_delete.append(rb)
+        else:
+            remaining_role_bindings.append(rb)
+
+    log.info("Deleting all resources that don't match the PMR.")
+    delete_many(client, role_bindings_to_delete, logger=log)
+
+    return remaining_role_bindings
+
+
+def create_rolebindings_for_profile_contributors(
+    client: Client,
+    profile: classes.Profile,
+    existing_rolebindings: List[RoleBinding],
+):
+    """Create RoleBindings for all contributors defined in a Profile, in the PMR.
+
+    If a RoleBinding already exists for the specific Contributor name and role, then
+    no API requests will happen.
+
+    Args:
+        client: The lightkube client to use.
+        profile: The PMR to iterate over its Contributors.
+        existing_rolebindings: List of existing RoleBindings, to avoid doing redundant
+                               API requests
+    """
+    existing_contributor_roles = kfam_resources_list_to_roles_dict(existing_rolebindings)
+
+    for contributor in profile.contributors or []:
+        if contributor.role not in existing_contributor_roles.get(contributor.name, []):
+            log.info("Will create RoleBinding for Contributor: %s", contributor)
+            client.apply(generate_contributor_rolebinding(contributor, profile.name))
```

Comment:
There should be some error handling here, as we know there could be some ApiError raised

Reply:
Hmm in case of an exception it should actually be something that would make the charm in blocked status as the admin should interfere with the existing resources. So the error should just be propagated.

Is there an exception that you had in mind that the code should catch and then proceed, or handle in a specific way?

Reply:
If it's something that should be handled by the consumer (charm in this case), then I think we should at least document it in the docstrings, just to make the API clearer.

Reply:
Ah good point!
367bc26